### PR TITLE
Implement a dedicated reactor-netty 1.0 instrumentation

### PR DESCRIPTION
**Pull request by mateuszrzeszutek**
This PR introduces a completely new reactor-netty 1.0 instrumentation; one that does not rely on netty client instrumentation to produce telemetry. The HTTP client span is now the parent of RESOLVE/CONNECT/SSL spans, as I suggested in #4617:

> In case of "normal"/higher-level HTTP clients I think it'd make perfect sense for them to be nested under the HTTP CLIENT span.

(Now I only need to change the kind of those nested spans from `INTERNAL` to `CLIENT`...)

CC @lmolkova

### CODE REVIEW

File: instrumentation/reactor-netty/reactor-netty-1.0/javaagent/src/test/groovy/io/opentelemetry/javaagent/instrumentation/reactornetty/v1_0/ReactorNettyClientSslTest.groovy [hunk 1]
```diff
@@ -54,37 +54,45 @@ class ReactorNettyClientSslTest extends AgentInstrumentationSpecification {
...
           errorEvent(thrownException.class, thrownException.message)
         }
         span(1) {
+          name "HTTP GET"
+          kind CLIENT
+          childOf span(0)
+          status ERROR
+          // netty swallows the exception, it doesn't make any sense to hard-code the message
+          errorEventWithAnyMessage(SSLHandshakeException)
+        }
```
Comment (trask) [inline instrumentation/reactor-netty/reactor-netty-1.0/javaagent/src/test/groovy/io/opentelemetry/javaagent/instrumentation/reactornetty/v1_0/ReactorNettyClientSslTest.groovy:70]: can you add attribute verification? (and below)

File: instrumentation/reactor-netty/reactor-netty-1.0/javaagent/src/test/groovy/io/opentelemetry/javaagent/instrumentation/reactornetty/v1_0/ReactorNettyConnectionSpanTest.groovy [hunk 2]
```diff
@@ -57,35 +57,35 @@ class ReactorNettyConnectionSpanTest extends InstrumentationSpecification implem
           hasNoParent()
         }
         span(1) {
+          name "HTTP GET"
+          kind CLIENT
+          childOf span(0)
+        }
```
Comment (trask) [inline instrumentation/reactor-netty/reactor-netty-1.0/javaagent/src/test/groovy/io/opentelemetry/javaagent/instrumentation/reactornetty/v1_0/ReactorNettyConnectionSpanTest.groovy:63]: same here (and below)

File: instrumentation/reactor-netty/reactor-netty-1.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/reactornetty/v1_0/ResponseReceiverInstrumentation.java [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+        @Advice.Local("otelCallDepth") CallDepth callDepth,
+        @Advice.Enter HttpClient.ResponseReceiver<?> modifiedReceiver,
+        @Advice.Return(readOnly = false) Mono<HttpClientResponse> returnValue) {
+
+      if (modifiedReceiver != null) {
+        returnValue = modifiedReceiver.response();
+      }
+
+      // needs to be called after original method to prevent StackOverflowError
+      callDepth.decrementAndGet();
```
Comment (trask) [inline instrumentation/reactor-netty/reactor-netty-1.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/reactornetty/v1_0/ResponseReceiverInstrumentation.java:99]: can you move the callDepth update to beginning of these methods? I'm just paranoid about resetting thread local state 😄

**Review by trask — APPROVED** [review]
nice 🎉

File: instrumentation/reactor-netty/reactor-netty-1.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/reactornetty/v1_0/HttpResponseReceiverInstrumenter.java [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+          () -> {
+            Context parentContext = Context.current();
+            contextHolder.parentContext = parentContext;
+            if (!instrumenter().shouldStart(parentContext, config)) {
+              return mono.contextWrite(ctx -> ctx.put(CLIENT_PARENT_CONTEXT_KEY, parentContext));
+            }
+
+            Context context = instrumenter().start(parentContext, config);
+            contextHolder.context = context;
+            return ContextPropagationOperator.runWithContext(mono, context)
```
Comment (lmolkova) [inline instrumentation/reactor-netty/reactor-netty-1.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/reactornetty/v1_0/HttpResponseReceiverInstrumenter.java:88]: nice! it looks like pure context API is still useful even without extra convenience we discussed to start spans

File: instrumentation/reactor-netty/reactor-netty-1.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/reactornetty/v1_0/DecoratorFunctions.java [hunk 5]
```diff
@@ -65,16 +71,27 @@ public void accept(M message, Throwable throwable) {
...
-      }
+  private static Context getChannelContext(
+      ContextView contextView, PropagatedContext propagatedContext) {
+    Context context = null;
+    if (propagatedContext.useClientContext) {
+      context = contextView.getOrDefault(ReactorContextKeys.CLIENT_CONTEXT_KEY, null);
+    }
+    if (context == null) {
+      context = contextView.getOrDefault(ReactorContextKeys.CLIENT_PARENT_CONTEXT_KEY, null);
+    }
```
[the remark this answers is not in the document]
Comment (mateuszrzeszutek) [reply instrumentation/reactor-netty/reactor-netty-1.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/reactornetty/v1_0/DecoratorFunctions.java:82]: It is -- when reactor-netty span is suppressed and client context is not accessible from the reactor's `ContextView` we still want to execute the `doOn*()` callbacks with the parent context in scope.

File: instrumentation/reactor-netty/reactor-netty-1.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/reactornetty/v1_0/HttpClientInstrumentation.java [hunk 6]
```diff
@@ -70,36 +64,22 @@ public void transform(TypeTransformer transformer) {
...
         @Advice.Argument(value = 0, readOnly = false)
-            BiConsumer<? super HttpClientRequest, ? super Connection> callback) {
+            BiConsumer<? super HttpClientRequest, ? super Connection> callback,
+        @Advice.Origin("#m") String methodName) {
+
       if (DecoratorFunctions.shouldDecorate(callback.getClass())) {
-        callback = new DecoratorFunctions.OnMessageDecorator<>(callback);
+        // use client context after request is sent, parent context before that
+        PropagatedContext propagatedContext =
+            "doAfterRequest".equals(methodName)
```
[the remark this answers is not in the document]
Comment (mateuszrzeszutek) [reply instrumentation/reactor-netty/reactor-netty-1.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/reactornetty/v1_0/HttpClientInstrumentation.java:79]: Good point, done.

File: instrumentation/reactor-netty/reactor-netty-1.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/reactornetty/v1_0/ResponseReceiverInstrumentation.java (hunk 3 above)
Comment (mateuszrzeszutek) [reply instrumentation/reactor-netty/reactor-netty-1.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/reactornetty/v1_0/ResponseReceiverInstrumentation.java]: I have to decrement the count _after_ the original `response*()` method is called, that's why they're the last thing I'm calling. I've wrapped those invocations in try-finally blocks - now the counter will always be decremented.

File: instrumentation/reactor-netty/reactor-netty-1.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/reactornetty/v1_0/HttpResponseReceiverInstrumenter.java (hunk 4 above)
Comment (mateuszrzeszutek) [reply instrumentation/reactor-netty/reactor-netty-1.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/reactornetty/v1_0/HttpResponseReceiverInstrumenter.java:91]: Yes, it turned out to be super useful this time 🚀

File: instrumentation/reactor-netty/reactor-netty-1.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/reactornetty/v1_0/ResponseReceiverInstrumentation.java (hunk 3 above)
Comment (trask) [reply instrumentation/reactor-netty/reactor-netty-1.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/reactornetty/v1_0/ResponseReceiverInstrumentation.java]: oh, i completely missed the `skipOn` and what was happening here 👍

**Pull request merged**